### ISSUE TITLE
Documentation for CHMOD and CHOWN

### ISSUE THREAD

**Issue by ZerooCool** [PrivateBin/PrivateBin#643]
Official : https://github.com/PrivateBin/PrivateBin/wiki/FAQ#what-are-the-recommended-file-and-folder-permissions-for-privatebin

My renderer : https://privatebin.visionduweb.fr/?c6137072beb310af#BFbSqT547nXrjFLd4JhtapTCCVh68XAJ1mGZmvKuizsh

My question : 

1- Create the storage folder if it is positioned elsewhere than the default location.
I wonder if this procedure is not incomplete, because, it would miss the scripts that exist by default in the data directory.
mkdir -p "$pbdata"

2- No information is present in the documentation, concerning the directory of the site itself. Can you confirm that this is correct?
The site container folder in /var/www/ will have CHMOD rights in 550 and will belong to "root: www-data".
dr-xr-x--- 13 root www-data 4096 juin  21 04:31 privatebin.visionduweb.fr

Either way, the tool works that way.
Thank you.

**Comment by elrido** [comment]
1. All of the contents in the data directory get created by privatebin automatically (as long as the PHP process has write access). If you move it somewhere else, after the installation has been running, the permissions need to be preserved in the move.
2. As the documentation explains, anything other then the data directory needs to be read-able. In [one version we did also convert the configuration file format](https://privatebin.info/news/v1.1.1-release.html) and it also needed to be write-able during this.

If your instance works, you did it correctly. Making files outside of the data folder read-only for the webserver and PHP process increases your security, but isn't required for the function of the tool. Imagine if there were an exploit in your webserver that would let an attacker create a file in the privatebin directory - having it read-only to the user the webserver runs as mitigates the attack.

Regarding checking your particular permissions: I am afraid that you will have to investigate this yourself. It depends on what users the webserver and php processes are running under and in what groups they are in. Given that your service works, they both seem to be members of the www-data group and so having the group permissions set to read and execute for the folder would be as is recommended.

**Comment by ZerooCool** [comment]
I have found this, in the documentation, but, i don't have the link.
I have modify the domain.

```
#!/bin/sh
# Les répertoires de PrivateBin :
pbpath='/var/www/privatebin.visionduweb.fr'
pbdata='/var/www/privatebin.visionduweb.fr/data'
# Utilisateur PHP de PrivateBin et du serveur web Apache (Se sont souvent les mêmes.)
# pbuser='php'
pbuser='www-data'
htgroup='www-data'
rootuser='root'

# Create the storage folder if it is positioned elsewhere than the default location.
# I wonder if this procedure is not incomplete, because, it would miss the scripts that exist by default in the data directory.
# mkdir -p "$pbdata"
# Une issue est ouverte : https://github.com/PrivateBin/PrivateBin/issues/643

echo "chmod"
find "${pbpath}/" -type f -print0 | xargs -0 chmod 0640
find "${pbpath}/" -type d -print0 | xargs -0 chmod 0550
find "${pbdata}/" -type f -print0 | xargs -0 chmod 0640
find "${pbdata}/" -type d -print0 | xargs -0 chmod 0750

echo "chown"
chown -R ${rootuser}:${htgroup} "${pbpath}/"
chown -R ${pbuser}:${rootuser} "${pbdata}/"
```

1- Create the storage folder if it is positioned elsewhere than the default location.
This procedure is not complete, because, it would miss the scripts that exist by default in the data directory. Then, the command ...
`mkdir -p "$pbdata"`
... is insufficient ... because ... In this folder, we have 3 files : 

purge_limiter.php  traffic_limiter.php salt.php

**Comment by ZerooCool** [comment]
There is the link : https://github.com/PrivateBin/PrivateBin/wiki/FAQ#what-are-the-recommended-file-and-folder-permissions-for-privatebin

**Comment by elrido** [comment]
These three files are not present, when you install PrivateBin and are generated by it, as I explained above. You don't even need to create the data directory, it will also be generated by PrivateBin, if not present.

The reason the script (which is provided by another user and not maintained by us) does the mkdir is so that they can then make the folder that usually is just above read-only. If you would do that before the folder is created, then PrivateBin could no longer create the folder by itself. I would therefore recommend to leave the mkdir in the script, just to be sure that it works on new installations that don't have the folder yet.

Important: The salt.php contains a key that is, among other things, used to generate the delete keys. Don't publish it or use one that you downloaded somewhere else. Let PrivateBin generate a unique one for you. With that salt one can generate valid delete links for every paste ID on your instance. If you ever change it, all delete keys previously generated are no longer valid.

**Comment by ZerooCool** [comment]
You tell me the folder below is read-only, so are you talking about the site folder?
So my site folder would be / var / www / mysite /
And so, should the "mysite" folder be read-only?
That is to say, in chmod 440?

**Comment by elrido** [comment]
Ideally, yes, the folder containing the "mysite", would be read-only. But remember that folders also need to be executable in order for applications to be allowed to list their contents or traverse them to access their contents, hence the script does:

```shell
find "${pbpath}/" -type f -print0 | xargs -0 chmod 0640
find "${pbpath}/" -type d -print0 | xargs -0 chmod 0550
```
So directories (type d) get 0550 and are readable and traversable and the files (type f) get 0640 - these could be restricted to 0440, as they don't need to be written to.

**Comment by ZerooCool** [comment]
I understand for 440 files.
I understand less for the 550 files when you tell me ideally, just readable for the file? 440? If I put in 440, I get a nice error 403;)
Well, for my part, the 550 suits me.
I asked the question, but, by re-reading the script, indeed, the file received rights 550.
So I will do so.

To conclude, if I understood correctly, the creation of the data folder, even, allows to modify the rights, and, to put it in read only.
This will therefore prevent the use of the tool to create.
On the other hand, the tool will still be able to read, in the case of saved messages, re-imported.
That's right ?
In the normal situation, data have 750 then we can publish a new message.

**Comment by ZerooCool** [comment]
I tested the script during my new deployment, and, indeed, it works well, to create the data folder.
So I corrected my wiki, and, I keep creating the data directory.

On the other hand, I had an alert, when I have CHMOD in 640 on the data folder, since no file exists.
It told me it was missing parameters.

Apparently, this does not matter, since the rights are correctly applied.
At the time, I believed that the script did not work, but, the message should simply mean that nothing happened in 640 because nothing existed then in data, during a brand new installation.
